### PR TITLE
Utilize NotificationService when we encounter git problems

### PR DESCRIPTION
**Pull request by taquitos**
- fix for #329 (If git repo is unreachable, exception is thrown and kills the server)
- Common exception handling for all git things
- Handle connection issues
- Raise all errors we don't explicitly handle
- Update notification service for extra non-user-fiendly notification details (like exception message)

### CODE REVIEW

File: app/shared/models/git_repo.rb [hunk 1]
```diff
@@ -106,32 +118,55 @@ def initialize(git_config: nil, local_folder: nil, provider_credential: nil, asy
...
         sleep(2)
         now = Time.now.utc
       end
 
       raise "Unable to start git repo #{git_config.git_url} in #{sync_setup_timeout_seconds} seconds" if now > sleep_timeout
-      logger.debug("Done starting up repo: #{self.git_config.git_url}")
+      logger.debug("Done starting up repo: #{git_config.git_url}")
+    end
+
+    def handle_exception(ex, message: nil)
```
Comment (KrauseFx) [inline app/shared/models/git_repo.rb]: Would be good to have documentation here on why we add `message` as a separate parameter

File: app/shared/models/git_repo.rb [hunk 2]
```diff
@@ -106,32 +118,55 @@ def initialize(git_config: nil, local_folder: nil, provider_credential: nil, asy
...
+      raise ex unless notification_service
+
+      message = ex.message.to_s
+      if message.contains("unable to access")
+        priority = Notification::PRIORITIES[:urgent]
+        notification_service.create_notification!(
+          priority: priority,
+          name: "Repo access error",
+          message: "Unable to acccess #{git_config.git_url}",
+          details: ex.message
```
Comment (KrauseFx) [inline app/shared/models/git_repo.rb]: We don't show `message` anywhere, except for in the CI terminal output?

File: app/shared/models/git_repo.rb [hunk 3]
```diff
@@ -106,32 +118,55 @@ def initialize(git_config: nil, local_folder: nil, provider_credential: nil, asy
...
       end
 
-      logger.debug("Adding task to setup repo #{self.git_config.git_url} at: #{local_folder}")
+      logger.debug("Adding task to setup repo #{git_config.git_url} at: #{local_folder}")
 
       setup_task = git_action_with_queue(ensure_block: proc { callback_block(async_start) }) do
-        logger.debug("Starting setup_repo #{self.git_config.git_url}".freeze)
+        logger.debug("Starting setup_repo #{git_config.git_url}".freeze)
         setup_repo
-        logger.debug("Done setup_repo #{self.git_config.git_url}".freeze)
```
Comment (KrauseFx) [inline app/shared/models/git_repo.rb]: why change from `self.` to non `self.`

Comment (AndrewMcBurney) [reply app/shared/models/git_repo.rb]: We removed the `RedundantSelf` cop in the #462 PR.

File: app/shared/models/git_repo.rb [hunk 4]
```diff
@@ -466,17 +545,22 @@ def clone_synchronously(repo_auth: self.repo_auth)
...
 
       self.temporary_storage_path = setup_auth(repo_auth: repo_auth)
       logger.debug("[#{git_config.id}]: Cloning git repo #{git_config.git_url} to #{@local_folder}")
-      Git.clone(git_config.git_url,
-                "", # checkout into the local_folder
-                path: local_folder,
-                recursive: true)
+
+      begin
+        Git.clone(git_config.git_url,
```
Comment (AndrewMcBurney) [inline app/shared/models/git_repo.rb]: This is a nit, but could we potentially style it as this instead:

```ruby
Git.clone(
  git_config.git_url,
  "", # checkout into the local_folder
  path: local_folder,
  recursive: true
)  
```

Thoughts?

File: app/shared/models/notification.rb [hunk 5]
```diff
@@ -69,13 +74,15 @@ class Notification
     # @param  [String] user_id
     # @param  [String] name
     # @param  [String] message
-    def initialize(id: nil, priority: nil, type: nil, user_id: nil, name: nil, message: nil, created_at: nil, updated_at: nil)
+    # @param  [String] details
+    def initialize(id: nil, priority: nil, type: nil, user_id: nil, name: nil, message: nil, created_at: nil, updated_at: nil, details: nil)
```
Comment (AndrewMcBurney) [inline app/shared/models/notification.rb]: Do you think we could change the rubocop max line length from [370](https://github.com/fastlane/ci/blob/master/.rubocop.yml#L174) to something like 120?

File: app/shared/models/git_repo.rb (hunk 2 above)
Comment (taquitos) [reply app/shared/models/git_repo.rb]: that's right, I renamed it

File: app/shared/models/git_repo.rb (hunk 4 above)
Comment (taquitos) [reply app/shared/models/git_repo.rb]: I agree 🌮

File: app/shared/models/notification.rb (hunk 5 above)
Comment (taquitos) [reply app/shared/models/notification.rb]: Yeah, totally. I think 120 is perfect 👍

Comment (taquitos) [reply app/shared/models/notification.rb]: (feel free to open an issue for that!)

Comment (AndrewMcBurney) [reply app/shared/models/notification.rb]: Sounds good! Will do!

Comment (AndrewMcBurney) [reply app/shared/models/notification.rb]: Updated in #506

File: app/shared/models/git_repo.rb (hunk 3 above)
Comment (KrauseFx) [reply app/shared/models/git_repo.rb]: Nice, sorry I missed that :)

**Pull request merged**